water_neuron/src/state.rs

```diff
@@ -500,6 +525,49 @@ impl State {
         );
     }
 
+    pub fn record_dispatch_icp_rewards(
+        &mut self,
+        neuron_6m_icp_amount: ICP,
+        sns_gov_amount: ICP,
+        timestamp: u64,
+        from_neuron_type: NeuronOrigin,
+    ) {
+        self.previous_week_fee_metrics.push_back(FeeMetrics {
+            revenue: sns_gov_amount,
+            reward: ICP::from_e8s(sns_gov_amount.0 + neuron_6m_icp_amount.0),
+            ts_secs: timestamp / SEC_NANOS,
+        });
+
+        loop {
+            if let Some(fee_metrics) = self.previous_week_fee_metrics.front() {
+                if timestamp_nanos() / SEC_NANOS - fee_metrics.ts_secs > ONE_WEEK_SECONDS {
```

Comment:
Nit: we expect `timestamp_nanos` to be always greater or equal to `fee_metrics.ts_sec`, but just to be on the safe side we could avoid subtraction to completely rule out underflows due to `u64`:
```
if timestamp_nanos() / SEC_NANOS  > fee_metrics.ts_secs + ONE_WEEK_SECONDS
```